homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,294 @@
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._is_on
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._available
+
+    @callback
+    def _availability_callback(self, availability):
+        """Update availability state."""
+        self._available = availability
+        self.async_schedule_update_ha_state()
+
+    async def async_added_to_hass(self):
+        """Register update callback."""
+        async_dispatcher_connect(self.hass, SIGNAL_AVAILABILITY,
+                                 self._availability_callback)
+
+        # Process the initial event now that the entity is created
+        if self._initial_event:
+            self.handle_event_callback(self._initial_event)
+
+
+class SW16Command(SW16Device):
+    """Singleton class to make HLK-SW16 command interface available to entities.
+
+    This class is to be inherited by every Entity class that is actionable
+    (switches/lights). It exposes the HLK-S16 command interface for these
+    entities.
+
+    The HLK-SW16 interface is managed as a class level and set during setup
+    (and reset on reconnect).
+    """
+
+    _protocol = None
+
+    @classmethod
+    def set_hlk_sw16_protocol(cls, protocol):
+        """Set the HLK-S16 asyncio protocol as a class variable."""
+        cls._protocol = protocol
+
+    @classmethod
+    def is_connected(cls):
+        """Return connection status."""
+        return bool(cls._protocol)
+
+    async def _async_handle_command(self, command, *args):
+        """Do bookkeeping for command, send it to HLK-SW16 and update state."""
+        if not self.is_connected():
+            raise HomeAssistantError('Cannot send command, not connected!')
+
+        if command == 'turn_on':
+            state = await self._protocol.turn_on(codecs.decode(
+                self._device_port.rjust(2, '0'), 'hex'))
+            _LOGGER.debug("new state:")
+            _LOGGER.debug(state)
+            self._is_on = True
+            # Update state of entity
+            await self.async_update_ha_state()
+
+        elif command == 'turn_off':
+            state = await self._protocol.turn_off(codecs.decode(
+                self._device_port.rjust(2, '0'), 'hex'))
+            _LOGGER.debug("new state:")
+            _LOGGER.debug(state)
+            self._is_on = False
+            # Update state of entity
+            await self.async_update_ha_state()
+
+        elif command == 'status':
+            state = await self._protocol.status()
+            _LOGGER.debug("new state:")
+            _LOGGER.debug(state)
+            # if the state is unknown or false, it gets set as true
+            # if the state is true, it gets set as false
+            self._is_on = state[self._device_port]
+
+
+class SwitchableSW16Device(SW16Command):
+    """HLK-SW16 entity which can switch on/off (eg: light, switch)."""
+
+    def _handle_event(self, event):
```

Comment:
This doesn't seem to be called anywhere?

Reply:
Yeah, I'm still working on wiring that up to receive state change events from the device protocol.

Reply:
removed this since wasn't needed